prog/mutation.go

```diff
@@ -269,6 +269,117 @@ func Minimize(p0 *Prog, callIndex0 int, pred func(*Prog, int) bool) (*Prog, int)
 		p0 = p
 		callIndex0 = callIndex
 	}
+
+	triedPaths := make(map[string]bool)
+
+	var rec func(p *Prog, call *Call, arg *Arg, path string) bool
+	rec = func(p *Prog, call *Call, arg *Arg, path string) bool {
+		switch typ := arg.Type.(type) {
+		case *sys.StructType:
+			for _, innerArg := range arg.Inner {
+				innerPath := fmt.Sprintf("%v-%v", path, innerArg.Type.Name())
+				if rec(p, call, innerArg, innerPath) {
+					return true
+				}
+			}
+		case *sys.UnionType:
+			if rec(p, call, arg.Option, fmt.Sprintf("%v-%v", path, arg.Option.Type.Name())) {
+				return true
+			}
+		case *sys.PtrType:
+			// TODO: remove optional ptrs
+			if arg.Res != nil {
+				return rec(p, call, arg.Res, fmt.Sprintf("%v-inner", path))
+			}
+			return false
+		case *sys.ArrayType:
+			for i, innerArg := range arg.Inner {
+				innerPath := fmt.Sprintf("%v-%v", path, i)
+				if (typ.Kind == sys.ArrayRangeLen && len(arg.Inner) > int(typ.RangeBegin)) ||
+				   (typ.Kind == sys.ArrayRandLen) {
+					if !triedPaths[innerPath] {
+						copy(arg.Inner[i:], arg.Inner[i+1:])
+						arg.Inner = arg.Inner[:len(arg.Inner)-1]
+						p.removeArg(call, innerArg)
+						assignSizesCall(call)
+
+						if pred(p, callIndex0) {
+							p0 = p
+						} else {
+							triedPaths[innerPath] = true
+						}
+
+						return true
+					}
+				}
+				if rec(p, call, innerArg, innerPath) {
+					return true
+				}
+			}
+		case *sys.IntType, *sys.FlagsType:
+			// TODO: try to set separate flags
+			if triedPaths[path] {
+				return false
+			}
+			if arg.Val == 0 {
+				triedPaths[path] = true
+				return true
+			}
+			arg.Val = 0
+			if pred(p, callIndex0) {
+				p0 = p
+			}
+			triedPaths[path] = true
+			return true
```

Comment:
If pred returns false, we will reclone the program and walk over arguments just to restore the int variable. I would prefer if we do just:
```go
			v0 := arg.Val
			arg.Val = 0
			if pred(p, callIndex0) {
				p0 = p
				return true
			} else {
				arg.Val = v0
				return false
			}
```